### PR TITLE
[Ray DAG API] Implement experimental ray DAG API for task/class 

### LINKED ISSUE #22066: Commit ray DAG API in experimental folder

### PR DESCRIPTION
**Pull request by jiaodong**
There're some polishing left regarding how to handle `ActorMethodNode` in progress, but other parts are good for review.
Many original implementations comes from Eric's prototype branch that this PR included in the branch history. Closes #22066 

## Diff summary

Notable changes:

All DAGNode instances are the result of calling `._bind` on a ray decorated func or class as IR. It shouldn't be directly called, constructed, used or manipulated by user in current setup. 

- Added DAGNode parent class that covers the following node types:
   - TaskNode: For function, takes (args, kwargs, options) it uses in execution
   - ActorNode: For class, takes (args, kwargs, options) it uses to run constructor
   - ActorMethodNode: For class method, in addition it takes its parent ActorNode as well as previous method call to reconstruct state needed, given guaranteed submission & execution order.
- DAGNode is able to recursively apply a given function for all DAGNodes it has, with methods introduced to find and replace such nodes.
   - In case of deeply nested args, it uses cloudpickle to find and replace all children DAGNodes.
   - For serve in particular we can enforce JSON serialization via a custom validation function on deploy() or build().

## Test plan

Unit tests for task / actor that covers various binding, nested args, options.

## Follow-up

- Revisit how we handle ActorMethodNode and its use of parsing `args` -- we likely have a better way.
- Revisit _PyObjScanner regarding threadsafety and serialization.
- Start extending serve implementation on deployment classes to make it work for serve pipeline. 

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/experimental/dag/actor_node.py
Comment (jiaodong) [inline python/ray/experimental/dag/actor_node.py:71]: We probably have a better way to handle this. It will be applied on next around of commits on this PR.

File: python/ray/experimental/dag/actor_node.py
Comment (jiaodong) [inline python/ray/experimental/dag/actor_node.py:83]: We probably have a better way to handle this. It will be applied on next around of commits on this PR.

File: python/ray/actor.py
Comment (ericl) [inline python/ray/actor.py:626]: Could we keep this as experimental for now?

File: python/ray/actor.py
Comment (ericl) [inline python/ray/actor.py:1012]: Experimental?

File: python/ray/experimental/dag/py_obj_scanner.py
Comment (ericl) [inline python/ray/experimental/dag/py_obj_scanner.py:43]: For thread-safety, I think we should assign `_cur[uuid] = self`, and delete `_cur[uuid]` in a finally block. Otherwise multiple threads could overwrite each others instances

File: python/ray/remote_function.py
Comment (ericl) [inline python/ray/remote_function.py:253]: Experimental.

File: python/ray/remote_function.py
Comment (ericl) [inline python/ray/remote_function.py:466]:
```suggestion
        change.
```

File: python/ray/remote_function.py
Comment (ericl) [inline python/ray/remote_function.py:256]:
```suggestion
                to changes.
```

File: python/ray/actor.py
Comment (jiaodong) [reply python/ray/actor.py:626]: sure, i didn't find "experimental" in `annotations.py` tho, what's the best practice in ray to mark it as experimental ?

Comment (ericl) [reply python/ray/actor.py:626]: How about not annotating it and marking it `**Experimental**` in docstring?

Comment (jiaodong) [reply python/ray/actor.py:626]: sure :)

File: python/ray/experimental/dag/py_obj_scanner.py
Comment (jiaodong) [reply python/ray/experimental/dag/py_obj_scanner.py:43]: good call -- for uuid here it's just some newly generated id for isolation, not related to the stable_uuid of a specific DAGNode right ?

Comment (ericl) [reply python/ray/experimental/dag/py_obj_scanner.py:43]: Yep just uuid.uuid4() would work, it's just a throwaway random number.

Comment (jiaodong) [reply python/ray/experimental/dag/py_obj_scanner.py:43]: I'm honestly still a bit lost about why `_cur` works here as class variable or cloudpickle in general but really want to make sure it's crystal clear .. let's chat about it tomorrow :)

File: python/ray/experimental/dag/py_obj_scanner.py
Comment (jiaodong) [inline python/ray/experimental/dag/py_obj_scanner.py:51]: @ericl one more question about this method, I think `reducer_override` is only introduced in python 3.8, https://docs.python.org/3/library/pickle.html#pickle.Pickler.reducer_override but we need to support at least python 3.6 and 3.7 wheels

**Comment by edoakes** [comment]
@jiaodong is this in a good state for me to review or should I hold off until other comments are addressed?

**Comment by jiaodong** [comment]
@edoakes i would say 90% of the content is good for review now given tests are passing for typical use cases. Two pieces that I think need some follow-ups in this PR are

- Revisit how we handle ActorMethodNode and its use of parsing args -- we likely have a better way.
- Revisit _PyObjScanner regarding threadsafety and serialization.

Which could happen in follow-up PR to address.

File: python/ray/experimental/dag/py_obj_scanner.py
Comment (ericl) [inline python/ray/experimental/dag/py_obj_scanner.py:54]:
```suggestion
            res = (lambda i: self._replace_index(i)), (index,)
```

I think this should work (not sure why I didn't do this in the first place).

File: python/ray/experimental/dag/dag_node.py
Comment (ericl) [inline python/ray/experimental/dag/dag_node.py:67]: Could you make the DAG traversal args private (_ leading)?

File: python/ray/experimental/dag/py_obj_scanner.py
Comment (jiaodong) [reply python/ray/experimental/dag/py_obj_scanner.py:54]: I got `TypeError: cannot pickle '_PyObjScanner' `object if i replace _PyObjScanner._cur with self .. after synced offline i will leave it to you to decide changes in this class

File: python/ray/experimental/dag/py_obj_scanner.py
Comment (ericl) [inline python/ray/experimental/dag/py_obj_scanner.py:41]: Unfortunately you have to explicitly register each subclass here.

File: python/ray/experimental/dag/dag_node.py
Comment (jiaodong) [inline python/ray/experimental/dag/dag_node.py:69]: this signature needs to be `def execute(self, *args, **kwargs)` to allow user to pass in runtime requests to execute rather than bound args. It's WIP in a follow-up PR.

File: python/ray/experimental/dag/actor_node.py
Comment (edoakes) [inline python/ray/experimental/dag/actor_node.py:8]:
```suggestion
    """Represents an actor creation in a Ray DAG."""
```

File: python/ray/experimental/dag/actor_node.py
Comment (edoakes) [inline python/ray/experimental/dag/actor_node.py:7]: As a general point, I feel it makes more sense for this to be a `ClassNode` because `Actor` is an implementation detail of the default way to execute the graph.

Comment (edoakes) [reply python/ray/experimental/dag/actor_node.py:7]: Other libraries, like Serve, will not use this to create an actor (rather a deployment in this case)

Comment (edoakes) [reply python/ray/experimental/dag/actor_node.py:7]: Same for TaskNode, should be FunctionNode IMO

File: python/ray/experimental/dag/actor_node.py
Comment (edoakes) [inline python/ray/experimental/dag/actor_node.py:11]:
```suggestion
        self._cls = cls
```

File: python/ray/experimental/dag/actor_node.py
Comment (edoakes) [inline python/ray/experimental/dag/actor_node.py:23]: Is this the default implementation if not passed to a different executor?

Comment (edoakes) [reply python/ray/experimental/dag/actor_node.py:23]: docstring please

File: python/ray/experimental/dag/actor_node.py
Comment (edoakes) [inline python/ray/experimental/dag/actor_node.py:38]: Is there a reason we do this lazily? Why not just raise an error right here?

File: python/ray/experimental/dag/dag_node.py
Comment (edoakes) [inline python/ray/experimental/dag/dag_node.py:101]: Is it valid to pass a node into `bound_options`? If there's no requirement for it, I'd suggest we don't support this for now to avoid added complexity

File: python/ray/experimental/dag/dag_node.py
Comment (edoakes) [inline python/ray/experimental/dag/dag_node.py:125]: this should share code with `self._get_all_child_nodes`

File: python/ray/experimental/dag/py_obj_scanner.py
Comment (edoakes) [inline python/ray/experimental/dag/py_obj_scanner.py:56]: I don't think I'm quite following this -- are we somehow forcing cloudpickle to do the replacement for us? The part I find most strange is that this method has side effects (you set `self._replace_table` and don't unset it). It also seems to make this not thread safe (which we need to make an explicit limitation if it's the case).

File: python/ray/experimental/dag/py_obj_scanner.py
Comment (edoakes) [inline python/ray/experimental/dag/py_obj_scanner.py:69]: what does this do? o.O

File: python/ray/experimental/dag/py_obj_scanner.py
Comment (edoakes) [inline python/ray/experimental/dag/py_obj_scanner.py:64]: we should not reach into private fields of `_PyObjScanner`, if this is necessary make it a public interface.

File: python/ray/experimental/dag/task_node.py
Comment (edoakes) [inline python/ray/experimental/dag/task_node.py:8]: Same as ActorNode, this should be `FunctionNode` IMO

File: python/ray/experimental/dag/dag_node.py
Comment (edoakes) [reply python/ray/experimental/dag/dag_node.py:101]: Looks like it's untested as well, let's either explicitly support & test it or not.

**Review by edoakes — COMMENTED** [review]
Some comments on code structure, two high-level comments:

- I don't understand how the serialization hackery is working (and left some comments about the interface). Let's please clearly document the behavior & limitations of this.
- I think we should rename `TaskNode` -> `FunctionNode` and `ActorNode` -> `ClassNode` given that this DAG will not necessarily be implemented as tasks/actors at all.

File: python/ray/experimental/dag/dag_node.py
Comment (jiaodong) [reply python/ray/experimental/dag/dag_node.py:101]: for bound_options, no, it's just runtime configurations and we can remove it.

File: python/ray/experimental/dag/actor_node.py
Comment (ericl) [reply python/ray/experimental/dag/actor_node.py:38]: The error is raised immediately on line 37, it's not lazy.

File: python/ray/experimental/dag/py_obj_scanner.py
Comment (ericl) [reply python/ray/experimental/dag/py_obj_scanner.py:56]: Yep, the easiest way to execute replace is via the deserialization hook of pickler. Note that it's thread safe per instance of PyObjScanner, which should be fine.

File: python/ray/experimental/dag/py_obj_scanner.py
Comment (ericl) [reply python/ray/experimental/dag/py_obj_scanner.py:69]: We need to workaround needing to reference instances of the pickler in the deserialization hook.

File: python/ray/experimental/dag/py_obj_scanner.py
Comment (ericl) [reply python/ray/experimental/dag/py_obj_scanner.py:64]: This is a private method call being generated from a private method, so it's fine.

File: python/ray/experimental/dag/actor_node.py
Comment (jiaodong) [reply python/ray/experimental/dag/actor_node.py:23]: yep each subclass of DAGNode implements its own executor by calling ray.remote()

File: python/ray/experimental/dag/dag_node.py
Comment (jiaodong) [reply python/ray/experimental/dag/dag_node.py]: i just gave it a try but find it slightly less clean given the current setup. So `_get_all_child_nodes` is only used in testing with the assumption that it always creates a new scanner object, but in replace_all_nodes we want to ensure we always use the same scanner for both find & replace.

File: python/ray/experimental/dag/py_obj_scanner.py
Comment (edoakes) [reply python/ray/experimental/dag/py_obj_scanner.py:66]: Oh, I didn't realize this is from the same class...

File: python/ray/experimental/dag/py_obj_scanner.py
Comment (edoakes) [reply python/ray/experimental/dag/py_obj_scanner.py:71]: @jiaodong add docstring?

File: python/ray/experimental/dag/test_actor_dag.py
Comment (edoakes) [inline python/ray/experimental/dag/test_actor_dag.py:1]: maybe I missed it, but I don't see a test where we pass a `ClassNode` into the constructor of another -- can you add that? Also one where it gets passed to a task.

File: python/ray/experimental/dag/dag_node.py
Comment (edoakes) [reply python/ray/experimental/dag/dag_node.py]: At a minimum you can probably do:
```python
def _find_bound_nodes(self, scanner: _PyObjScanner):
    return s.find_nodes(
            [self._bound_args, self._bound_kwargs, self._bound_options]
        )
```

**Review by edoakes — APPROVED** [review]
LGTM to merge as a first cut, @jiaodong please file issues for follow-ups in the pipelines milestone

File: python/ray/experimental/dag/test_actor_dag.py
Comment (jiaodong) [reply python/ray/experimental/dag/test_actor_dag.py]: added and it passed as expected :)

File: python/ray/experimental/dag/dag_node.py
Comment (jiaodong) [reply python/ray/experimental/dag/dag_node.py]: yep i've tried that but makes some tests a bit more ugly and breaks a few assumptions it had regarding shared scanner :/ Let me put it as follow up item to clean up these functions with tests all together.

**Review by ericl — APPROVED** [review]

**Comment by jiaodong** [comment]
Only failed test was runtime env test due to timeout, should be unrelated. Let me retry a few times to see if we can make it green.

**Comment by BalaBalaYi** [comment]
@jiaodong @ericl 
Hi. May i ask, has the community considered integrating the 'DAG API' with 'streaming'. If so, what is the relationship between 'DAG API' and 'DataStream API'?

By the way, we've done this('DAG API' + 'streaming') before in a similar way(we call it 'Tide API'). And it was also verified in our production.  @ashione

**Comment by BalaBalaYi** [comment]
I can't show u the whole doc because of the company security issue.

Here are some general info:

for functions:
1. support DAG and DCG in massive paral
2. data consistency when failover(exactly once)
3. static and dynamic scheduling
...


for API and example(a training case):
```
@ray.tide.remote
class Worker(object):
    
    @ray.tide.initializer
    def init(self, context):
        self.__data = load_data(context)
        

    @ray.tide.processor
    def process(self, model, context):
        model2 = calc_model(model, self.__data)
        context.emit(model2)

@ray.tide.remote
class Merger(object):
    def __init__(self):
        self.__started = False
        self.__model_list = []

    @ray.tide.generator
    def generate(self, context):
        if not self.__started:
            self.__started = True
                context.emit(emtpy_model())

    @ray.tide.processor
    def process(self, model, context):
        self.__model_list.append(model)
        if len(self.__model_list) == context.vertex('worker').para:
            merged_model = merge_model(self.__model_list)
            context.emit(merged_model)
            self.__model_list = []

    @ray.tide.checkpoint.getter
    def save_cp(self)
        return [self.__started, self.__model_list]

    @ray.tide.checkpoint.setter
    def load_cp(self, data)
        self.__started, self.__model_list = data


from ray.tide.shuffle_func import broadcast
with ray.tide.Session(name="demo") as session:
    session.vertex("merger", Merger, 1)
        .to("worker", Worker, 10, shuffle_func=broadcast)
        .to("merger").run()
    # other normal ray driver code 
    ...
```

**Pull request merged**
